docs/COREDUMP_PACKAGE_METADATA.md

```diff
@@ -0,0 +1,127 @@
+---
+title: Package Metadata for Core Files
+category: Interfaces
+layout: default
+---
+
+# Package Metadata for Core Files
+
+*Intended audience: hackers working on userspace subsystems that create ELF binaries
+or parse ELF core files.*
+
+## Motivation
+
+ELF binaries get stamped with a unique, build-time generated hex string identifier called
+`build-id`, which gets embedded as an ELF note called `.note.gnu.build-id`.
+In most cases, this allows to associated a stripped binary with its debugging symbols.
+It is used, for example, to dynamically fetch DWARF symbols from a debuginfo server, or
+to query the local package manager and find out the package metadata or, again, the DWARF
+symbols.
+
+However, there are a number of popular use cases where this is insufficient. Think of the
+use-case of systemd-coredump running on a host and collecting cores from applications
+running inside containers of any sort. Unless the host and the containers are running the
+same distribution, which is increasingly unlikely, the host's configured debuginfo will
+not have a record of the binary. And even if the distribution is the same, the binary
+might not necessarily be from the distribution itself, but might be supplied by the
+owner of the container.
+There are also cases where the host runs isolated or air-gapped, and thus it is (intentionally)
+unable to reach a remote server.
+
+When reaching out is not possible, the packaging system local configuration state will
+hold more information - which package the binary came from, the version, and so on.
+But again in the case of a container this only works if there is a match.
+And even if the host and guest match, this relies on the host actually having a package
+manager local state configured, which is increasingly not the case, for example in case
+of minimalistic read-only images that are built and updated by a distro, but as a
+monolithic image that is updated as a single artifact, without any local packaging state
+to save space and memory.
+
+For all these cases, adding additional metadata to a binary at build time might be
+desirable, so that systemd-coredump and other services analyzing core files are able to
+extract said metadata simply from the core file itself, without having to rely on
+external services or tools which might not be available.
+
+## Implementation
+
+This document will attempt to define a common metadata format specification, so that
+multiple implementers might use it when building packages, or core file analyzers, and
+so on.
+
+The metadata will be embedded in a single, new ELF header section, in a key-value JSON
+format. Implementers working on parsing core files should take care in not assuming
```

Comment:
should take care in not assuming … → should not assume a specific list list of keys